docs/RationaleSeq.md

```diff
@@ -0,0 +1,89 @@
+  as a "latch". Examples: "gated SR latch", "D latch".
+  - Edge-triggered **flip-flop:** An edge-sensitive memory element. Captures
+  the input value on one or both clock edges. Variants: posedge FF, negedge
+  FF, "edge-sensitive" FF (captures the input value on both edges),
+  resettable FF.
+- Abstract models:
+  - **Register:** A synchronous, resettable memory element. Can be
+  implemented using any of the above "circuit level" elements.
+
+## The register operation
+
+The `seq.reg` op models an abstract notion of a "register", independent of
+its implementation (e.g. latch, D flip-flop). It is intended to be 'lowered'
+to a specific implementation, which may be an `sv.always_ff` block, a device
+primitive instantiation, or even a feedback loop in all `comb` logic. It has
+three operands and one attribute:
+
+- **input**: The value to be captured at 'clock'. Generally called 'q'.
+Accepts any type, results in the same type.
+- **clock**: Capture 'value' on the positive edge of this signal.
+- **reset**: Signal to set the state to 'resetValue'. Optional.
+- **resetValue**: A value which the state is set to upon reset. Optional.
+
+```mlir
+%d = seq.reg %input, %clk [, %reset [%resetValue] ] : $type(input)
+```
+
+### Rationale
+
+Several design decisions were made in defining this op:
+
+- Inclusion of 'reset' signal: the most common lowering case is expected to
+result in a SystemVerilog `always_ff`, which _should_ include an `if (rst)`
+block. This operand makes lowering to that form easier. Omission of if would
+require some (potentially complex) analysis to find the reset mux.
+- Inclusion of 'resetValue' attribute: if we have a 'reset' signal, we need
+to include a value to avoid complex inversion tricks.
+- Omission of 'enable': enables are easily modeled via a multiplexer on the
+input with one of the muxs inputs as the output of the register. This
+property makes 'enables' easier to detect than reset in the common case.
+- Omission of 'negedge' event on 'clock': this is easily modeled by inverting the
+clock value.
+- Omission of 'dual edge' event on 'clock': this is not terribly common. If
+we find a compelling use case, we could either create a new op for "dual
+edge" register, or create a "clock double" op which would interpose on the
+clock value. (And be detected by a lowering step).
+- Omission of edge conditions on 'reset': The reset style is generally a
+design-wide choice and (if done correctly) shouldn't affect correctness. It
+should, therefore, be determined by a lowering pass.
+
+### Future considerations
+
+- Enable signal: if this proves difficult to detect (or non-performant if we
+do not detect and generate the SystemVerilog correctly), we can build it into
+the reg op.
+- Reset style and clock style: how should we model posedge vs negedge clocks?
+Async vs sync resets? There are some reasonble options here: attributes on
+the `reg` op or `clock` and `reset` types which are parameterized with that
```

Comment:
Curious if you are thinking to propagate the reset type from the signal connected to the reg.reset

Reply:
The idea behind `clock` and `reset` types is that we'd be able to store information about their uses -- posedge/negedge, async/sync. This is -- apparently -- useful for generating the clock/reset trees. It would also allow us to merely check that the clock/reset types match before doing some merging operation -- like creating a shift register from a series of regs. (Not that we couldn't do that by checking some attributes, it'd just be more code.)